stdlib/universe/holt_winters/holt_winters.go

```diff
@@ -123,47 +99,55 @@ func (r *HoltWinters) Do(vs *array.Float64) *array.Float64 {
 	if r.seasonal {
 		size += m
 	}
-	r.tracker.track(size)
-	defer r.tracker.free(size)
-	params := make([]float64, 6, size)
-	params[4] = l0
-	params[5] = b0
+	params := mutable.NewFloat64Array(r.alloc)
+	defer params.Release()
+	params.Resize(size)
+	params.Set(4, l0)
+	params.Set(5, b0)
 	if r.seasonal {
 		for i := 0; i < m; i++ {
 			if vs.IsValid(i) {
-				params = append(params, vs.Value(i)/l0)
+				params.Set(i+6, vs.Value(i)/l0)
 			} else {
-				params = append(params, 0)
+				params.Set(i+6, 0)
 			}
 		}
 	}
 
 	// Determine best fit for the various parameters
 	minSSE := math.Inf(1)
-	var bestParams []float64
+	var bestParams *mutable.Float64Array
 	for alpha := hwGuessLower; alpha < hwGuessUpper; alpha += hwGuessStep {
 		for beta := hwGuessLower; beta < hwGuessUpper; beta += hwGuessStep {
 			for gamma := hwGuessLower; gamma < hwGuessUpper; gamma += hwGuessStep {
 				for phi := hwGuessLower; phi < hwGuessUpper; phi += hwGuessStep {
-					params[0] = alpha
-					params[1] = beta
-					params[2] = gamma
-					params[3] = phi
+					params.Set(0, alpha)
+					params.Set(1, beta)
+					params.Set(2, gamma)
+					params.Set(3, phi)
 					sse, params := r.optim.Optimize(r.sse, params, r.epsilon, 1)
```

Comment:
This is the line that confused me. In general I'd like to write the code such that the calls to `New` and `Release` are in easy to spot pairs.

This line hides the fact that the Optimizer creates a new array and reassigns that on top of the params var and thus requiring a call to Release. At first read it looks like params is being updated inplace but not reallocated. 

I see two ways we can clarify this code.

1. Rewrite Optimizer so that it doesn't reallocate params rather does inplace update
2. Move this inner loop into a function so `defer params.Release()` can be used. This is one is likely harder as we don't want to release the bestParams ...

The first option should be a small change to update the input parameters and not return them as a new array.

Reply:
I see your point 👍 

I have already tried the second option and that was pretty hard and I couldn't get to a working solution.

The problem with the first one is that the Optimizer shouldn't have side effects on `params`. That first array should be left untouched, as it is used as it is but with `a, b, c, d` changed every time.

I think that code is a bit tricky, because when you `params :=` you are not resetting params, but shadowing it.

Reply:
In other terms, the _outer_ params must leave for the whole `for` loop, while the `params` created by `Optimize` should be released conditionally.

Reply:
@affo Why can't the Optimizer method mutate the input params? I looked over the code and it seemed fine to change that at first glance.

Reply:
I think that the initial `params` must stay the same at every iteration in the `for` loop (they get only changes on the first 4 values). The optimizer only reads from them to generate a new solution that can be the best or not. Anyways, it must start from those initial params every time, it cannot change them. Otherwise it would produce incorrect results

Reply:
ok, then lets add some comments to make this clear what is going on, i.e. Optimize creates a new params and the initial params is used several times. Perhaps we could rename the outer params to initParams so it is clear that it is being reused instead of shadowing?

Reply:
Definitely agree 👍